attributed_text/src/error.rs

```diff
@@ -0,0 +1,219 @@
+        match self.kind {
+            ErrorKind::InvalidBounds => write!(
+                f,
+                "range {}..{} out of bounds for len {}",
+                self.start, self.end, self.len
+            ),
+            ErrorKind::InvalidRange => {
+                write!(f, "invalid range {}..{}: start > end", self.start, self.end)
+            }
+            ErrorKind::NotOnCharBoundary => {
+                if let Some(b) = self.boundary {
+                    let which = match b.which {
+                        Endpoint::Start => "start",
+                        Endpoint::End => "end",
+                    };
+                    write!(
+                        f,
+                        "range {}..{}: {} index {} not on UTF-8 boundary (char {}..{})",
+                        self.start, self.end, which, b.index, b.char_start, b.char_end
+                    )
+                } else {
+                    write!(
+                        f,
+                        "range {}..{} not on UTF-8 boundary",
+                        self.start, self.end
+                    )
+                }
+            }
+        }
+    }
+}
+
+impl core::error::Error for Error {}
+
+/// The non-exhaustive category of an error.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// Provided range indices were out of bounds relative to the text length.
+    InvalidBounds,
+
+    /// The provided range had `start > end`.
+    InvalidRange,
+
+    /// Either `start` or `end` was not aligned to a UTF-8 character boundary.
+    NotOnCharBoundary,
+}
+
+/// Identifies which endpoint of a range failed boundary validation.
+///
+/// This type is surfaced via [`BoundaryInfo`], which is attached to [`Error`]
+/// for boundary-related failures.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Endpoint {
+    /// The `start` endpoint of the range.
+    Start,
+
+    /// The `end` endpoint of the range.
+    End,
+}
+
+/// Details about an offending index that was not on a UTF-8 character boundary.
+///
+/// Returned by [`Error::boundary`] when the error kind is
+/// [`ErrorKind::NotOnCharBoundary`].
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct BoundaryInfo {
+    /// Which endpoint (`start` or `end`) was invalid.
+    pub which: Endpoint,
+
+    /// The offending byte index.
+    pub index: usize,
+
+    /// The start byte index of the enclosing UTF-8 codepoint.
+    pub char_start: usize,
+
+    /// The end byte index (exclusive) of the enclosing UTF-8 codepoint.
+    pub char_end: usize,
+}
+
+fn enclosing_char_span<T: TextStorage>(text: &T, index: usize) -> Option<(usize, usize)> {
+    let len = text.len();
+    if index > len {
+        return None;
+    }
+    if text.is_char_boundary(index) {
+        return Some((index, index));
+    }
+
+    // Previous boundary (max 3 bytes back)
+    let mut s = index;
+    for _ in 0..4 {
+        if s == 0 {
+            break;
+        }
```

Comment:
In principle this should be impossible, as `text.is_char_boundary(0)` should always be `true`.

```suggestion
        // We can never wrap around `0` here, as when `s == 0` before the decrement,
        // the previous call to `text.is_char_boundary` (either in this loop or above)
        // will have returned `true`.
```